File: depth_fisrt_search/depth_fisrt_search.py
class Node:
  def __init__(self, name):
    self.children = []
    self.name = name
  
  def add_child(self, name):
    self.children.append(Node(name))

  def depth_first_search(self, array):
    array.append(self.name)
    for child in self.children:
      child.depth_first_search(array)
    return array

root = Node("A")
root.add_child("B")
root.add_child("C")
root.children[0].add_child("D")  # B -> D
root.children[0].add_child("E")  # B -> E
root.children[1].add_child("F")  # C -> F

result = []
result = root.depth_first_search(result)
print(result)  # Output: ['A', 'B', 'D', 'E', 'C', 'F']

File: depth_fisrt_search/depth_fisrt_search.ts
class MyNode {
    name: string;
    children: MyNode[] = [];

    constructor(name: string) {
        this.name = name;
    }

    addChild(name: string): MyNode {
        const child = new MyNode(name);
        this.children.push(child);
        return this;
    }

    depthFirstSearch(array: string[]): string[] {
        array.push(this.name);
        for (const child of this.children) {
            child.depthFirstSearch(array);
        }
        return array;
    }
}

const root = new MyNode("A");
root.addChild("B").addChild("C");
root.children[0].addChild("D");

console.log(root.depthFirstSearch([]));
